parse_type.go

```diff
@@ -0,0 +1,419 @@
+package typegen
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"unicode"
+)
+
+// TypeExpr is the root of the AST. It represents a Go type expression that can be:
+//   - A named type (with an optional package alias/path).
+//   - A pointer to another TypeExpr.
+//   - A slice or array of another TypeExpr.
+//   - A map of (KeyExpr -> ValueExpr).
+//   - A generic instantiation: BaseTypeExpr[T1, T2, ...]  (the base can itself be a named type).
+type TypeExpr struct {
+	// If non-empty, we have a pointer depth.
+	// e.g. **[]Foo => PtrDepth=2, then an underlying that is "[]Foo"
+	PtrDepth int
+
+	// If ArrayLen >= 0, that means it's an array (fixed length). -1 means not an array.
+	ArrayLen int
+
+	// If IsSlice is true, we have a slice. (We do not store the length for slices.)
+	IsSlice bool
+
+	// If IsMap is true, then MapKey/MapValue are valid.
+	IsMap    bool
+	MapKey   *TypeExpr
+	MapValue *TypeExpr
+
+	// Base package path (or alias) and name, if this is a named type.
+	// E.g. PkgPath="github.com/foo/bar", Name="MyType".
+	// If the underlying is something else (map, slice, etc.), these may be "".
+	PkgPath string
+	Name    string // base type name, e.g. "CborLink"
+
+	// If we have generic parameters, they are stored here. For example:
+	//   "CborLink[*SimpleTypeTwo, SomeOther]" => Generics has 2 entries.
+	// If we do not have generics, Generics is nil or empty.
+	Generics []*TypeExpr
+
+	// If we have a parent in the parse, it can be stored for convenience, but not required.
+}
+
+// parseTypeName parses a type expression (like reflect.Type.String()) into a *TypeExpr.
+func parseTypeName(full string) (*TypeExpr, error) {
+	// We’ll produce tokens, then parse them.
+	tokens, err := tokenizeTypeString(full)
+	if err != nil {
+		return nil, fmt.Errorf("tokenize failed: %w", err)
+	}
+	p := &typeParser{tokens: tokens}
+	expr, err := p.parseTypeExpr()
+	if err != nil {
+		return nil, fmt.Errorf("parseTypeExpr failed near token %d (\"%s\"): %w",
+			p.pos, p.currToken(), err)
+	}
+	// We expect no trailing tokens.
+	if p.pos < len(tokens) {
+		return nil, fmt.Errorf("extra tokens after parse: %v", tokens[p.pos:])
+	}
+	return expr, nil
+}
+
+// tokenizeTypeString splits things like "map[string]*github.com/foo/bar.Baz" into
+// tokens: "map", "[", "string", "]", "*", "github.com/foo/bar.Baz" etc.
+// We keep bracket depth logic minimal. We'll handle identifiers, punctuation, brackets, etc.
+func tokenizeTypeString(s string) ([]string, error) {
+	var tokens []string
+	i := 0
+	for i < len(s) {
+		c := s[i]
+		switch c {
+		case '[', ']', ',', '*', '(', ')':
+			tokens = append(tokens, string(c))
+			i++
+		case ' ':
+			// skip
+			i++
+		case '{', '}':
+			// If reflect ever gives curly braces (uncommon), we’d parse them.
+			// For now, let’s just treat them as punctuation tokens if needed.
+			tokens = append(tokens, string(c))
+			i++
+		default:
```

Comment:
In this case, I'd consider dropping all the "is letter" logic, etc. and instead taking until the next punctuation. We're not trying to validate the input so we might as well be permissive.